### DISCUSSION TITLE
[A-L] (12 gennaio 2023) Esercizio 4

### DISCUSSION THREAD

**Discussion by Frisbon** [sapienzastudentsnetwork/progettazione-di-algoritmi#74] [Esami A-L]
_tengo a precisare che sulla traccia originale ha scritto 2016, si è sbagliato, è del 2023._
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/a2453fd1-47b3-4889-858a-9bf840439e74)

**Comment by Frisbon** [comment]
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/06c9a335-5d85-47cb-8225-34a214f2120f)

![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/fafa51b1-670d-478a-a913-807f15b3dbd6)

```python
def Risparmia(SOL,VIS,COSTR,G,nodo): #visita solo se esistono colori alternati
    VIS[nodo] = 1;
    for adiacente in nodo:
        if VIS[adiacente] == 0 and adiacente.colore != nodo.colore:
            SOL.add(( nodo, adiacente) ) #aggiungi arco
            Risparmia( SOL, VIS, G, adiacente)
        else:
            COSTR.add(nodo) #aggiungo il nodo costretto
    return VIS
    

def Costretto(SOL,VIS,G,nodo): # visita qualsiasi cosa, se costretto
    VIS[nodo] = 1;
    for adiacente in nodo:
        if VIS[adiacente] == 0:
            SOL.add(( nodo, adiacente) ) #aggiungi arco
            VIS = Risparmia( SOL, VIS, G, adiacente) #ricomincia a "risparmiare"
    return VIS

def es(G,nodo):
    VIS = [0,...,0]; SOL = {}; COSTR = {}
    while 0 in VIS: #fino a quando non visito tutti
        VIS = Risparmia(SOL,VIS,G,nodo)
        for x in COSTR:
            VIS = Costretto(SOL,VIS,G,x)
```

**Reply by Frisbon** [reply]
Nota, ho trovato che Exyss ha negli appunti una soluzione molto ma molto più corta....

**Comment by cornflexxx** [comment]
Praticamente identico a #93 l'unica differenza è che qui aggiungiamo per primi gli archi $(u,v)$ tali che  $C(u) \neq C(v)$.

**Comment by daromacarlo** [comment]
```c
//algoritmo greedy
//MODIFICO L'ALGORITMO DI PRIM
def albero_colorato(G){
        sol = []
	vis = [0,...,0]
	while exists w in V(G) | vis[w] == 0{
		if exists (v,u) = arco in E(G) | (v.colore != u.colore && vis[u] == 0 && vis[v] == 1){
			sol.append(v,u)
		}
		else{
			(v,u) = arco in E(G)| (vis[u] == 0 && vis[v] == 1)
			sol.append(v,u)
		}
	}
}
return sol

/* 
costo finale = O(nm) 
   dato che il ciclo while è ripetuto m volte mentre le operazioni interne al ciclo costano O(2m).
*/
```

**Comment by CiottoloMaggico** [comment]
Algoritmo di Prim che utilizza la funzione w per il peso degli archi
![immagine](https://github.com/user-attachments/assets/731f8874-c0a8-4967-b77f-8d1307499ce5)

**Comment by niccolo-fato** [comment]
![Screenshot 2025-05-24 alle 21 26 22](https://github.com/user-attachments/assets/64307d21-b072-4ccc-a9d9-e526fa5ca248)
